ambry-frontend/src/main/java/com.github.ambry.frontend/BufferCopyAsyncWritableChannel.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2018 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package com.github.ambry.frontend;
+
+import com.github.ambry.rest.RestServiceErrorCode;
+import com.github.ambry.rest.RestServiceException;
+import com.github.ambry.router.AsyncWritableChannel;
+import com.github.ambry.router.Callback;
+import com.github.ambry.router.FutureResult;
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.io.SequenceInputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicLong;
+
+
+/**
+ * An implementation of {@link AsyncWritableChannel} that will copy all received chunks into a sequence of byte arrays.
+ * Meant for interacting with blocking content consumers that would otherwise require a background thread.
```

Comment:
I don't understand this comment, how is a background thread is required without this?

Also, we call `future.done()` in `write()`, is this implementation still async? 

Would read/write race condition be a problem for callers? 

We can chat offline.

Reply:
I think this is still nonblocking since `future.done()` only counts down on a latch but does not wait or have any synchronized segments. It does perform a buffer copy within the method, though, so the time spent in this method is proportional to the size of `src`.

There should not be a read/write race condition as long as callers call the `getContentAsInputStream()` method from within the `readInto()` callback; the stream should contain all chunks written. If they were to call it before then, the stream will at least contain the chunks written at call time (it may, but is not guaranteed to, contain some chunks written after the `iterator()` method was called.

From [javadocs](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentLinkedQueue.html#iterator()): The iterator "guarantees to traverse elements as they existed upon construction of the iterator, and may (but is not guaranteed to) reflect any modifications subsequent to construction"